server/etcdserver/corrupt.go

```diff
@@ -258,57 +259,157 @@ func (cm *corruptionChecker) CompactHashCheck() {
 			cm.mux.Unlock()
-			cm.lg.Info("finished compaction hash check", zap.Int("number-of-hashes-checked", i+1))
-			return
+			return true
 		}
 		cm.lg.Warn("skipped revision in compaction hash check; was not able to check all peers",
 			zap.Int("number-of-peers-checked", peersChecked),
 			zap.Int("number-of-peers", len(peers)),
-			zap.Int64("revision", hash.Revision),
+			zap.Int64("revision", leaderHash.Revision),
+			zap.Int64("compactRevision", leaderHash.CompactRevision),
 		)
+		// The only case which needs to check next hash
+		return false
 	}
-	cm.lg.Info("finished compaction hash check", zap.Int("number-of-hashes-checked", len(hashes)))
-	return
+
+	// Detected hashes mismatch
+	// The first step is to figure out the majority with the same hash.
+	memberCnt := len(peers) + 1
+	quorum := memberCnt/2 + 1
+	quorumExist := false
+	for k, v := range hashMap {
+		if len(v) >= quorum {
+			quorumExist = true
+			// remove the majority, and we might raise alarms for the left members.
+			delete(hashMap, k)
+			break
+		}
+	}
+
+	if !quorumExist {
+		// If quorumExist doesn't exist, then only raise alarm for the least minority
+		cm.lg.Error("Detected compaction hash mismatch but can't identify the corrupted members, so only raise alarm for the least minority",
+			zap.String("leader-id", leaderId.String()),
+			zap.Int64("leader-revision", leaderHash.Revision),
+			zap.Int64("leader-compact-revision", leaderHash.CompactRevision),
+			zap.Uint32("leader-hash", leaderHash.Hash),
+		)
+
+		// find out the members which are the least minority
+		var (
+			minCnt    int = math.MaxInt
+			hashVal   uint32
+			memberIDs types.IDSlice
+		)
+
+		for k, v := range hashMap {
+			if v.Len() < minCnt {
+				minCnt = v.Len()
+				hashVal = k
+				memberIDs = v
+			}
+		}
+
+		for _, pid := range memberIDs {
+			cm.hasher.TriggerCorruptAlarm(pid)
+		}
+		delete(hashMap, hashVal)
+
+		cm.lg.Error("Detected compaction hash mismatch but can't identify the corrupted members, so only raise alarm for the least minority",
+			zap.String("leader-id", leaderId.String()),
+			zap.Int64("leader-revision", leaderHash.Revision),
+			zap.Int64("leader-compact-revision", leaderHash.CompactRevision),
+			zap.Uint32("leader-hash", leaderHash.Hash),
+			zap.Uint32("peer-hash", hashVal),
+			zap.String("peer-ids", memberIDs.String()),
+		)
+	}
+
+	// Raise alarm for the left members if the quorum is present.
+	// But we should always generate error log for debugging.
+	for k, v := range hashMap {
+		for _, pid := range v {
+			if quorumExist {
+				cm.hasher.TriggerCorruptAlarm(pid)
+			}
+		}
+		cm.lg.Error("Detected compaction hash mismatch",
+			zap.String("leader-id", leaderId.String()),
+			zap.Int64("leader-revision", leaderHash.Revision),
+			zap.Int64("leader-compact-revision", leaderHash.CompactRevision),
+			zap.Uint32("leader-hash", leaderHash.Hash),
+			zap.Uint32("peer-hash", k),
+			zap.String("peer-ids", v.String()),
+			zap.Bool("alarm-raised", quorumExist),
```

Comment:
This doesn't make sense, we always raise at this point in code even if `quorumExist = false`

Reply:
updated.